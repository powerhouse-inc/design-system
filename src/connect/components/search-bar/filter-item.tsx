import { mergeClassNameProps } from '@/powerhouse';

export type FilterItemType = {
    id: string;
    label: string;
    icon?: React.JSX.Element;
};

export interface FilterItemProps extends React.HTMLAttributes<HTMLDivElement> {
    item: FilterItemType;
}

export const FilterItem: React.FC<FilterItemProps> = props => {
    const { item, ...containerProps } = props;

    return (
        <div
            {...mergeClassNameProps(
                containerProps,
                'flex flex-row h-full items-center gap-x-4 justify-between px-2',
            )}
        >
            {item.icon}
            <div className="text-sm font-semibold text-slate-200">
                {item.label}
            </div>
        </div>
    );
};
