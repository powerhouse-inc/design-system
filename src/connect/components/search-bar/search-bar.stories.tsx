import { defaultFilterItems } from '@/connect';
import type { Meta, StoryObj } from '@storybook/react';
import { ConnectSearchBar } from './search-bar';

const meta: Meta<typeof ConnectSearchBar> = {
    title: 'Connect/Components/SearchBar',
    component: ConnectSearchBar,
    argTypes: {
        value: { control: { type: 'text' } },
        onChange: { control: { type: 'action' } },
        placeholder: { control: { type: 'text' } },
        filterLabel: { control: { type: 'text' } },
        filterItems: { control: { type: 'object' } },
        onFilterSelect: { control: { type: 'action' } },
        selectedFilter: {
            control: { type: 'select' },
            options: [undefined, ...defaultFilterItems.map(item => item.id)],
        },
    },
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Primary: Story = {
    args: {
        placeholder: 'Search Files',
        filterLabel: 'File type',
        filterItems: defaultFilterItems,
    },
};
