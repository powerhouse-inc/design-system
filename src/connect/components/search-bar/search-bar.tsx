import {
    DropdownMenu,
    Icon,
    TextInput,
    TextInputProps,
    mergeClassNameProps,
} from '@/powerhouse';
import { useMemo } from 'react';
import { FilterItem, FilterItemType } from './filter-item';

export interface ConnectSearchBarProps extends TextInputProps {
    placeholder?: string;
    filterLabel?: string;
    filterItems?: Array<FilterItemType>;
    selectedFilter?: string;
    onFilterSelect?: (filterId: string) => void;
}

export const ConnectSearchBar: React.FC<ConnectSearchBarProps> = props => {
    const {
        value,
        onChange,
        placeholder,
        filterLabel,
        filterItems,
        selectedFilter,
        onFilterSelect = () => {},
        ...containerProps
    } = props;

    const items = useMemo(
        () =>
            filterItems?.map(item => ({
                id: item.id,
                content: <FilterItem item={item} />,
            })),
        [filterItems],
    );

    const selectedItemFilter = filterItems?.find(
        item => item.id === selectedFilter,
    );

    const filterLabelContent = selectedItemFilter ? (
        <FilterItem item={selectedItemFilter} className="gap-x-1" />
    ) : (
        filterLabel && (
            <div className="mr-2 text-sm font-semibold text-[#6F767E]">
                {filterLabel}
            </div>
        )
    );

    const startAdornment = <Icon name="search" className="mr-3" />;

    const endAdornment = (
        <DropdownMenu
            items={items || []}
            onItemClick={onFilterSelect}
            popoverProps={{ placement: 'bottom right', offset: 4 }}
            className="ml-3 flex h-full flex-row items-center outline-none"
            menuClassName="bg-grey-50 border border-[#F4F4F4] rounded-xl p-2"
            menuItemClassName="hover:bg-[#F4F4F4] rounded-lg overflow-hidden cursor-pointer h-10"
        >
            {filterLabelContent}
            <Icon name="chevron-down" />
        </DropdownMenu>
    );

    return (
        <TextInput
            value={value}
            onChange={onChange}
            startAdornment={startAdornment}
            endAdornment={endAdornment}
            inputProps={{ placeholder, className: 'text-sm' }}
            {...mergeClassNameProps(
                containerProps,
                'text-[#6F767E] bg-grey-50 px-4 h-[52px] items-center rounded-xl border border-[#EFEFEF]',
            )}
        />
    );
};
