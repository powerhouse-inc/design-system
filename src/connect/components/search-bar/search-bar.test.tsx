import BarChartIcon from '@/assets/icons/bar-chart-fill.svg';
import BriefCaseIcon from '@/assets/icons/briefcase-fill.svg';
import GlobeIcon from '@/assets/icons/globe.svg';
import PersonIcon from '@/assets/icons/person-fill.svg';
import ProjectIcon from '@/assets/icons/project-fill.svg';
import { fireEvent, render, screen } from '@testing-library/react';
import { ConnectSearchBar, ConnectSearchBarProps } from './search-bar';

const filterItems: ConnectSearchBarProps['filterItems'] = [
    {
        id: 'project',
        label: '.project',
        icon: ProjectIcon,
    },
    {
        id: 'budget',
        label: '.budget',
        icon: BarChartIcon,
    },
    {
        id: 'profile',
        label: '.profile',
        icon: PersonIcon,
    },
    {
        id: 'legal',
        label: '.legal',
        icon: BriefCaseIcon,
    },
    {
        id: 'atlas',
        label: '.Atlas',
        icon: GlobeIcon,
    },
];

describe('ConnectSearchBar Component', () => {
    it('should match snapshot', () => {
        const { asFragment } = render(
            <ConnectSearchBar
                placeholder="Search Files"
                filterLabel="File type"
                filterItems={filterItems}
                textFieldProps={{ 'aria-label': 'search input' }}
            />,
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it('should open filter menu when click on filters', () => {
        render(
            <ConnectSearchBar
                placeholder="Search Files"
                filterLabel="File type"
                filterItems={filterItems}
                data-testid="search-bar"
                textFieldProps={{ 'aria-label': 'search input' }}
            />,
        );

        fireEvent.click(screen.getByText('File type'));

        expect(screen.getByText('.project')).toBeInTheDocument();
        expect(screen.getByText('.budget')).toBeInTheDocument();
        expect(screen.getByText('.profile')).toBeInTheDocument();
        expect(screen.getByText('.legal')).toBeInTheDocument();
        expect(screen.getByText('.Atlas')).toBeInTheDocument();
    });

    it('should call onFilterSelect when a filter is selected', () => {
        const onFilterSelect = jest.fn();

        render(
            <ConnectSearchBar
                placeholder="Search Files"
                filterLabel="File type"
                filterItems={filterItems}
                data-testid="search-bar"
                textFieldProps={{ 'aria-label': 'search input' }}
                onFilterSelect={onFilterSelect}
            />,
        );

        fireEvent.click(screen.getByText('File type'));
        fireEvent.click(screen.getByText('.project'));

        expect(onFilterSelect).toBeCalledTimes(1);
        expect(onFilterSelect).toBeCalledWith('project');
    });

    it('should call onChange when input value is changed', () => {
        const onChange = jest.fn();

        render(
            <ConnectSearchBar
                placeholder="Search Files"
                filterLabel="File type"
                filterItems={filterItems}
                data-testid="search-bar"
                textFieldProps={{ 'aria-label': 'search input' }}
                onChange={onChange}
            />,
        );

        fireEvent.change(screen.getByRole('textbox'), {
            target: { value: 'test' },
        });

        expect(onChange).toBeCalledTimes(1);
        expect(onChange).toBeCalledWith('test');
    });

    it('should display filter selected in filter label', () => {
        render(
            <ConnectSearchBar
                placeholder="Search Files"
                filterLabel="File type"
                filterItems={filterItems}
                data-testid="search-bar"
                textFieldProps={{ 'aria-label': 'search input' }}
                selectedFilter="project"
            />,
        );

        expect(screen.getByText('.project')).toBeInTheDocument();
        expect(screen.queryByText('File type')).not.toBeInTheDocument();
    });
});
