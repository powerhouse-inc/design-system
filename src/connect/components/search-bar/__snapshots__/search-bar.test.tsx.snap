// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConnectSearchBar Component > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-row text-[#6F767E] bg-grey-50 px-4 h-[52px] items-center rounded-xl border border-[#EFEFEF]"
  >
    <svg
      class="mr-3"
      style="width: 24px; height: 24px;"
    >
      <use
        href="/icons.svg#search"
      />
    </svg>
    <div
      class="flex flex-1 min-w-0"
      data-rac=""
    >
      <input
        aria-label="search input"
        class="flex flex-1 outline-none bg-inherit min-w-0 text-sm"
        data-rac=""
        id="react-aria-:r0:"
        placeholder="Search Files"
        title=""
        type="text"
        value=""
      />
    </div>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="menu"
      class="ml-3 flex h-full flex-row items-center outline-none"
      data-rac=""
      id="react-aria-:r5:"
      type="button"
    >
      <div
        class="mr-2 text-sm font-semibold text-[#6F767E]"
      >
        File type
      </div>
      <svg
        style="width: 24px; height: 24px;"
      >
        <use
          href="/icons.svg#chevron-down"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
